### PR TITLE
Enable Support for Updatable Layers

### PR DESCRIPTION
**Pull request by peterg17**
This PR includes the Kyrix front-end and back-end update handlers that allow for updates in layers if specified. It makes changes to many files in Kyrix but the update code is only executed if a Transform object is passed an update functions object such as in https://github.com/tracyhenry/Kyrix/blob/31b5d24cf471c052b49124b4a6a45c3070c31c19/compiler/examples/nba/transforms.js#L58. If this update functions object is passed instead of an array of column name strings, the transform is flagged as updatable and any layer it is passed to also flagged. Then the Kyrix front-end attaches an update option to the jump menu as seen below and a drag/drop handler if x/y attributes are given update functions.

Otherwise, the visualizations should behave as they do normally in Kyrix. I tested several of the main Kyrix visualizations to make sure that backwards compatibility is achieved, since I do not want this PR to break anything. Also, these changes were rebased onto the current master branch and should not have any merge conflicts.

# Update Menu Example and Drag and Drop Example

kyrix-dragging-success

hierarchical-kyrix-updates

# How to Run Updates Code

1. Build Kyrix source code by running `./run-kyrix.sh --build` from the project root folder

2. Make changes to NBA timeline visualization by specifying `transforms.teamTimelineTransformUpdates` or convert back to normal version by using `transforms.teamTimelineTransform` in nba visualization spec here: https://github.com/tracyhenry/Kyrix/blob/31b5d24cf471c052b49124b4a6a45c3070c31c19/compiler/examples/nba/nba.js#L35

3. Run NBA timeline visualization by running `./compile.sh nba.js` in the `./compiler/examples/nba/` folder (relative to the project root folder)

4. Test updatable USMap visualization by running  `./compile.sh election.js` in the `./compiler/examples/election/` folder (relative to the project root folder)

5. Test regular USMap visualization for backwards compatibility by running `./compile.sh USMap.js` in the `./compiler/examples/USMap/` folder (relative to the project root folder)

### CODE REVIEW

**Comment by peterg17** [comment]
The only bug that i've found from rebasing onto the new master branch to fast-forward changes is that it breaks my update menu. When the user clicks on an input box in the update menu to change it, the whole jump menu is removed. I've tried to checking if removePopover() is called in some unintended way, but haven't figured out why this happens yet....



https://user-images.githubusercontent.com/2456071/105553250-e411a980-5cb9-11eb-9eba-e89b6153efeb.mov

**Comment by tracyhenry** [comment]
Thanks Peter. Will take a look as soon as I can.

**Comment by tracyhenry** [comment]
@peterg17 the bug was due to this line of code: https://github.com/tracyhenry/Kyrix/blob/a623ac03f11c75028e76188b32721c7a7c93ff2d/front-end/js/pageOnLoad.js#L226
which removes popovers when there is a click to the "window". You'd need to stop event propagation for your input elements. Let me know if this is easy to do.

**Comment by peterg17** [comment]
Thanks for finding the bug, I an added an event handler to stop propagation on just the input elements. This seems to have fixed the bug and I will add the update example to the documentation now.

**Comment by tracyhenry** [comment]
> Thanks for finding the bug, I an added an event handler to stop propagation on just the input elements. This seems to have fixed the bug and I will add the update example to the documentation now.

great, thanks.

File: back-end/src/main/java/project/UpdateRequest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+
+/** Created by wenbo on 1/4/18. */
+public class UpdateRequest {
+
+    public UpdateRequest() {}
+
+    // private variables
```
Comment (tracyhenry) [inline back-end/src/main/java/project/UpdateRequest.java:11]: @peterg17 can you help clean up the code a little bit to remove comments and logs for debugging? You can leave a few commented logs there if you think they are helpful for future debugging purposes.

Comment (peterg17) [reply back-end/src/main/java/project/UpdateRequest.java:11]: Yes, I can definitely clean up the comments. Thanks for pointing this out, comments and logging should not be in the committed code unless its very helpful for debugging.

Comment (peterg17) [reply back-end/src/main/java/project/UpdateRequest.java]: I went through and cleaned up comments/logging in the files I edited or added, mostly in UpdateRequestHandler.java, UpdateRequest.java, and jump.js. I also re-ran the election USMap visualization and nba visualization to make sure they still worked fine.

Comment (tracyhenry) [reply back-end/src/main/java/project/UpdateRequest.java]: This is great. Thanks! Let me know when the readme is ready. 

Also as a heads up - it might take a little more work before we can merge the PR. I looked through the code and there is a bunch places that I'd like to improve, including the API and where some code should live. But it should all be straightforward code refactoring. I will help - let's work on this together.

Comment (peterg17) [reply back-end/src/main/java/project/UpdateRequest.java]: I wasn't sure whether to update the README or the wiki, so I updated the Kyrix API wiki page for now. I noted my changes to the Transform object in the wiki here (https://github.com/tracyhenry/Kyrix/wiki/API-Reference#transform-for-updatable-layer). Also, I'm working on a new page in the wiki to describe updatable visualizations in a more end-to-end way that gives context. I added a new page in the wiki here (https://github.com/tracyhenry/Kyrix/wiki/Updatable-Kyrix-Visualizations), which I am adding content to. 

Thanks for the heads up, that makes total sense! It's a large PR and I would happy to help where I can to improve the code. We can coordinate here or over email, whatever works for you 👍

Comment (tracyhenry) [reply back-end/src/main/java/project/UpdateRequest.java]: Yes it's fine to put documentation in the places you mentioned. For the transform documentation, is it possible to add that to the existing parameter list? And as you mentioned in the comment, it'll be helpful to just 

In any case, I'll do one pass of review today. And let's communicate on Github here since it's easier  to reference code.

Comment (peterg17) [reply back-end/src/main/java/project/UpdateRequest.java]: Ok, i'll change the edits I made to the API Reference wiki page!

And just to clarify, only attributes that are listed in the schema object are made updatable.  The developer only needs to specify the identity reverse function lots of times if they want to allow the user to edit all of those attributes and only needs to be done once. I'm not sure how much effort is being saved by making a default function since they would still have to import and specify that default function in the Transform.

Comment (tracyhenry) [reply back-end/src/main/java/project/UpdateRequest.java]: Yes letting the user import `identityFunction` makes perfect sense. One improvement would be just asking the user to specify updatable attributes, and default unspecified ones to non-updatable. Then in your example we don't need to specify null for attributes such as `home_team` and `away_team`.

File: compiler/src/Canvas.js [hunk 2]
```diff
@@ -46,6 +46,11 @@ function Canvas(id, w, h, wString, hString) {
 
 // add layer to a canvas
 function addLayer(layer) {
+    console.log(
+        `adding new layer in canvas: ${this.id} with id: ${this.layers.length}`
+    );
```
Comment (tracyhenry) [inline compiler/src/Canvas.js:51]: I think these log lines might not be necessary.

File: compiler/src/Canvas.js [hunk 3]
```diff
@@ -46,6 +46,11 @@ function Canvas(id, w, h, wString, hString) {
...
+    layer.id = this.id;
+    layer.canvasId = this.layers.length;
```
Comment (tracyhenry) [inline compiler/src/Canvas.js:53]: Is `layer.id` and `layer.canvasId` absolutely necessary?

File: compiler/src/template-api/addUSMap.js [hunk 4]
```diff
@@ -5,6 +5,14 @@ const Layer = require("../Layer").Layer;
 const Transform = require("../Transform").Transform;
 const getBodyStringOfFunction = require("./Utilities").getBodyStringOfFunction;
 
+const updateStateQuery = `SELECT cs.name, cs.state_id, cs.total_dem_votes, cs.total_rep_votes,
```
Comment (tracyhenry) [inline compiler/src/template-api/addUSMap.js:8]: These seem to be hardcoded for the election dataset. 

I think for simplicity, maybe we can make the usmap one in a separate PR? since it involves editing another set of APIs (the template APIs).

File: compiler/src/template-api/json-schema/USMap.json [hunk 5]
```diff
@@ -102,6 +102,10 @@
             "type": "string",
             "maxLength": 20,
             "default": "Rate"
+        },
+        "updatesEnabled": {
+            "enum": [true, false],
```
Comment (tracyhenry) [inline compiler/src/template-api/json-schema/USMap.json:107]: you can say `"type": "boolean"` here

File: front-end/index.html [hunk 6]
```diff
@@ -25,6 +25,8 @@
         <script src="https://cdn.jsdelivr.net/npm/gpu.js@2.0.0-rc.19/dist/gpu-browser.js"></script>
         <script src="https://cdn.jsdelivr.net/npm/jquery-validation@1.17.0/dist/jquery.validate.min.js"></script>
         <script src="https://d3js.org/d3.v5.min.js"></script>
+        <!-- <script src="https://unpkg.com/d3-simple-slider"></script> -->
```
Comment (tracyhenry) [inline front-end/index.html:28]: can we load this slider through a CDN?

File: compiler/src/Transform.js [hunk 7]
```diff
@@ -58,10 +58,31 @@ function Transform(
...
+                    funcBody = "return " + funcBody;
+                    this.reverseFunctions[colName] = funcBody;
+                }
+            }
+            this.allowUpdates = true;
+        } else {
+            throw new Error(
+                "Constructing Transform: columnNames must be either an Array of strings or an Object mapping string -> function"
+            );
+        }
```
Comment (tracyhenry) [inline compiler/src/Transform.js:85]: The object way of specifying a transform is still "experimental", so don't worry to support it. You can remove these for now.

File: compiler/src/Transform.js [hunk 8]
```diff
@@ -84,30 +105,58 @@ function Transform(
             console.log(this.query);
         }
         this.separable = query["separable"] || true;
+        this.dependencies = [];
```
Comment (tracyhenry) [inline compiler/src/Transform.js:108]: can you help explain what `dependencies` are for? This ideally should be in the documentation too.

File: compiler/src/Transform.js [hunk 9]
```diff
@@ -84,30 +105,58 @@ function Transform(
...
         return;
     }
 
-    if (typeof separable !== "boolean")
-        throw new Error("Constructing Transform: separable must be boolean.");
-    if (!Array.isArray(columnNames))
+    // same block of code as above, but for other case!
+    // TODO: remove duplicate code
+    if (Array.isArray(columnNames)) {
```
Comment (tracyhenry) [inline compiler/src/Transform.js:114]: The stuff for updates should be mixed with `columnNames` and should be put in a separate field (e.g. 'updates'). The reasons are that 1) `columnNames` is an array for specifying aliases and 2) putting in a separate field will allow the user to omit the specification of non-updatable fields.

**Review by tracyhenry — CHANGES_REQUESTED** [review]
Hey Peter, I took at look at the changes to the compiler and made some comments. Can you take a look and let me know if you'd like to discuss? 

For some questions/changes I could've figured out or change myself but I have a paper deadline on Wednesday... so apologize for not being "autonomous" enough.

File: compiler/src/Canvas.js (hunk 3 above)
Comment (peterg17) [reply compiler/src/Canvas.js:53]: Yes these are needed to specify the `dependencies` array, which is an array of layers that are dependent on the current layer, defined by their <canvasId, layerId> unique identifier pair. An example where this is used is in the hierarchical election map example. I should document the `dependencies` attribute as part of the Layer API, which you pointed out in another comment in the PR.

File: front-end/index.html (hunk 6 above)
Comment (peterg17) [reply front-end/index.html:28]: Actually, the slider is no longer needed so i can just delete this line!

File: compiler/src/Transform.js (hunk 8 above)
Comment (peterg17) [reply compiler/src/Transform.js:108]: Yep! I can put in the Layer API, since the dependencies array is the way to specify whether there is some hierarchical dependency between two (or more) layers. For example, in the hierarchical election example, there is a dependency between the main county dynamic layer and the main state dynamic layer.

File: compiler/src/Canvas.js (hunk 2 above)
Comment (peterg17) [reply compiler/src/Canvas.js]: Fixed!

File: compiler/src/template-api/json-schema/USMap.json (hunk 5 above)
Comment (peterg17) [reply compiler/src/template-api/json-schema/USMap.json]: Thanks, fixed!

File: compiler/src/Transform.js (hunk 9 above)
Comment (peterg17) [reply compiler/src/Transform.js:113]: Ok, that makes sense. I wanted to avoid having to change every time that a Transform is created in the current visualizations, but I think I can achieve this just using an optional `updateAttrs` parameter, which defaults to an empty object.

Comment (tracyhenry) [reply compiler/src/Transform.js:113]: Exactly. Let me know when you fix this part and I'll move on reviewing other parts of the code. Thanks!

**Comment by peterg17** [comment]
Closing this PR in favor of a smaller scoped PR which doesn't change the USMap template API, and focuses on basic updates ([https://github.com/tracyhenry/Kyrix/pull/176](https://github.com/tracyhenry/Kyrix/pull/176))
